// reactive-localstorage/es/localstorage.js
export class LocalStorage {
    constructor(window, storage) {
        this._events = {};
        if (window) {
            this.supported = true;
        }
        else {
            this.supported = false;
            window = new Window();
        }
        this.storage = storage || window.localStorage;
        listen(window, ({ key, newValue }) => {
            if (!key)
                return;
            this.feed(key, newValue);
        });
    }
    get length() {
        return this.storage.length;
    }
    key(index) {
        return this.storage.key(index);
    }
    getItem(key) {
        return this.storage.getItem(key);
    }
    setItem(key, value) {
        this.storage.setItem(key, value);
        this.emit('change', key, value);
    }
    removeItem(key) {
        this.storage.removeItem(key);
        this.emit('change', key, null);
    }
    clear() {
        this.storage.clear();
    }
    on(name, fn) {
        let set = this._events[name];
        if (!set) {
            this._events[name] = set = new Set();
        }
        set.add(fn);
    }
    off(name, fn) {
        const set = this._events[name];
        if (set) {
            set.delete(fn);
        }
    }
    feed(key, newValue) {
        this.emit('change', key, newValue);
    }
    emit(name, ...args) {
        const set = this._events[name];
        if (set) {
            for (const fn of set.values()) {
                fn.apply(null, args);
            }
        }
    }
}
export function listen(target, cb) {
    if (target.addEventListener) {
        target.addEventListener('storage', cb, true);
    }
    else if (target.attachEvent) {
        target.attachEvent('storage', cb);
    }
    else {
        return false;
    }
    return true;
}
