// mobx-localstorage/es/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { toJS, observable, action, createAtom } from 'mobx';
import localStorage from 'reactive-localstorage';
export class LocalStorage {
    constructor() {
        this[Symbol.toStringTag] = 'Map';
        this._keyBoxes = new Map();
        this._valueBoxes = new Map();
        this._keys = new Set();
        this._keysAtom = createAtom('mobx storage keys');
        this._is_all_keys_observed = false;
        localStorage.on('change', action((key, value) => {
            if (typeof key === 'string') {
                if (typeof value === 'string') {
                    this._set(key, value);
                }
                else {
                    this._delete(key);
                }
            }
            else {
                this._clear();
            }
        }));
    }
    get length() {
        return this.size;
    }
    get size() {
        this._keysAtom.reportObserved();
        this._syncAllKeys();
        return this._keys.size;
    }
    key(index) {
        return localStorage.key(index);
    }
    getItem(key) {
        return this.get(key);
    }
    setItem(key, value) {
        this.set(key, value);
    }
    removeItem(key) {
        this.delete(key);
    }
    has(key) {
        return this._get_key_box(key).get();
    }
    get(key) {
        return this._get_value_box(key).get();
    }
    set(key, value) {
        localStorage.setItem(key, JSON.stringify(toJS(value)));
        return this;
    }
    delete(key) {
        const has = (this._is_all_keys_observed
            ? this._keys.has(key)
            : typeof localStorage.getItem(key) === 'string');
        localStorage.removeItem(key);
        return has;
    }
    clear() {
        localStorage.clear();
    }
    keys() {
        this._keysAtom.reportObserved();
        this._syncAllKeys();
        return this._keys.values();
    }
    *values() {
        for (const key of this.keys()) {
            yield this._get_value_box(key).get();
        }
    }
    *entries() {
        for (const key of this.keys()) {
            yield [key, this._get_value_box(key).get()];
        }
    }
    [Symbol.iterator]() {
        return this.entries();
    }
    forEach(callbackfn, thisArg) {
        this._keysAtom.reportChanged();
        this._syncAllKeys();
        for (const [key, value] of this.entries()) {
            callbackfn.call(thisArg, value, key, this);
        }
    }
    _set(key, value) {
        this._keysAtom.reportChanged();
        this._keys.add(key);
        const keyBox = this._keyBoxes.get(key);
        if (keyBox) {
            keyBox.set(true);
        }
        const valueBox = this._valueBoxes.get(key);
        if (valueBox) {
            valueBox.set(parseValue(value));
        }
    }
    _delete(key) {
        this._keysAtom.reportChanged();
        this._keys.delete(key);
        const keyBox = this._keyBoxes.get(key);
        if (keyBox) {
            keyBox.set(false);
        }
        const valueBox = this._valueBoxes.get(key);
        if (valueBox) {
            valueBox.set(null);
        }
    }
    _clear() {
        this._keysAtom.reportChanged();
        this._keys.clear();
        for (const box of this._keyBoxes.values()) {
            box.set(false);
        }
        for (const box of this._valueBoxes.values()) {
            box.set(null);
        }
    }
    _get_key_box(key) {
        const boxes = this._keyBoxes;
        let box = boxes.get(key);
        if (!box) {
            const value = (this._is_all_keys_observed
                ? false
                : typeof localStorage.getItem(key) === 'string');
            boxes.set(key, box = observable.box(value));
        }
        return box;
    }
    _get_value_box(key) {
        const boxes = this._valueBoxes;
        let box = boxes.get(key);
        if (!box) {
            const value = this._is_all_keys_observed ? null : localStorage.getItem(key);
            boxes.set(key, box = observable.box(parseValue(value)));
        }
        return box;
    }
    _syncAllKeys() {
        if (this._is_all_keys_observed)
            return;
        this._is_all_keys_observed = true;
        const len = localStorage.length;
        const boxes = this._keyBoxes;
        const keys = this._keys;
        keys.clear();
        for (let i = 0; i < len; i++) {
            const key = localStorage.key(i);
            if (typeof key !== 'string')
                continue;
            keys.add(key);
            if (boxes.has(key))
                continue;
            boxes.set(key, observable.box(true));
        }
    }
}
__decorate([
    action,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", void 0)
], LocalStorage.prototype, "set", null);
__decorate([
    action,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], LocalStorage.prototype, "delete", null);
__decorate([
    action,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], LocalStorage.prototype, "clear", null);
export function parseValue(value) {
    try {
        return JSON.parse(value);
    }
    catch (e) {
        // ignore if cannot parse as JSON
    }
    return value;
}
export default new LocalStorage();
